### PR TITLE
8340491: Thread stack-base assertion should report which thread has the un-set stack

### PR DESCRIPTION
**Pull request by dholmes-ora**
Please review this simple enhancement to an assertion so we can identify which thread had the problem.

I had to move the function to the cpp file due to include file issues.

We are limited in what we can print due to this being used very early in the thread initialization process - in particular no ResourceMarks are possible so we can't print the name.

Testing:
 - tier 5 (used to debug [JDK-8340401](https://bugs.openjdk.org/browse/JDK-8340401))
 - tiers 1-3 sanity

Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340491](https://bugs.openjdk.org/browse/JDK-8340491): Thread stack-base assertion should report which thread has the un-set stack (**Enhancement** - P3)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) Review applies to [e6358a57](https://git.openjdk.org/jdk/pull/21102/files/e6358a57d0d73909a2c30c465edea45027c371d5)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21102/head:pull/21102` \
`$ git checkout pull/21102`

Update a local copy of the PR: \
`$ git checkout pull/21102` \
`$ git pull https://git.openjdk.org/jdk.git pull/21102/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21102`

View PR using the GUI difftool: \
`$ git pr show -t 21102`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21102.diff">https://git.openjdk.org/jdk/pull/21102.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21102#issuecomment-2362586549)

### CODE REVIEW

**Review by stefank — APPROVED** [review]
Looks fine, but could you explain when this is needed?

If it is the current thread that is crashes then we already print the thread info.

If it is crashing when called on another thread, it might be easier to add informative asserts in that sub-system instead. You could for example add a non-asserting getter and perform all relevant checks on the caller side.

Another alternative could be to leverage the VMErrorCallback functionality I introduced a little while ago. With that you can register a callback to be called by the hs_err printer if the current thread crashes. With that you can have much more context about what the calling code was doing and what information that thread has available. That would look something like this (~I've not compiled this, so there might be errors~ Edit: updated the code):

```
diff --git a/src/hotspot/share/nmt/memMapPrinter.cpp b/src/hotspot/share/nmt/memMapPrinter.cpp
index 5f920b102a9..a2aae39b0ca 100644
--- a/src/hotspot/share/nmt/memMapPrinter.cpp
+++ b/src/hotspot/share/nmt/memMapPrinter.cpp
@@ -44,6 +44,7 @@
 #include "runtime/vmThread.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/ostream.hpp"
+#include "utilities/vmError.hpp"

 // Note: throughout this code we will use the term "VMA" for OS system level memory mapping

@@ -164,8 +165,29 @@ class CachedNMTInformation : public VirtualMemoryWalker {

 /////// Thread information //////////////////////////

+class VMAInspectionErrorCallback : public VMErrorCallback {
+  const void* const   _from;
+  const void* const   _to;
+  const Thread* const _thread;
+
+public:
+  VMAInspectionErrorCallback(const void* from, const void* to, const Thread* thread) :
+    _from(from), _to(to), _thread(thread) {}
+
+  void call(outputStream* st) override {
+    st->print_cr("Crashing while printing VMA details for " PTR_FORMAT " " PTR_FORMAT " for thread %s with id: %d",
+                 p2i(_from),
+                 p2i(_to),
+                 _thread->name(),
+                 _thread->osthread() != nullptr ? _thread->osthread()->thread_id() : 0);
+  }
+};
+
 // Given a VMA [from, to) and a thread, check if vma intersects with thread stack
 static bool vma_touches_thread_stack(const void* from, const void* to, const Thread* t) {
+  VMAInspectionErrorCallback on_error(from, to, t);
+  VMErrorCallbackMark mark(&on_error);
+
   // Java thread stacks (and sometimes also other threads) have guard pages. Therefore they typically occupy
   // at least two distinct neighboring VMAs. Therefore we typically have a 1:n relationshipt between thread
   // stack and vma.
```

Note that in this context we do have a ResourceMark, so it should be fine to print the name.

**Comment by dholmes-ora** [comment]
Thanks for the review @stefank !

> If it is crashing when called on another thread, it might be easier to add informative asserts in that sub-system instead. You could for example add a non-asserting getter and perform all relevant checks on the caller side.

The assert is firing when `stack_base()` is called on another thread. In that case we had nothing to tell us which thread it was, hence this enhancement to provide that information. We want this asserting getter to be used, and for the assert to fire .

File: src/hotspot/share/runtime/thread.cpp
Comment (tstuefe) [inline src/hotspot/share/runtime/thread.cpp:155]: Can't we still have this as inlined version? Either move this to thread.inline.hpp or give us a release variant without assert in Thread body in thread.hpp.

Comment (shipilev) [reply src/hotspot/share/runtime/thread.cpp:157]: Yeah. I think we can "just" outline the checking method like:

```
thread.hpp:
  address stack_base() const           { assert_stack_base(); return _stack_base; }
  void assert_stack_base() const NOT_DEBUG_RETURN;

thread.cpp:
#ifdef ASSERT
void  Thread::assert_stack_base() const {
  // Note: can't report Thread::name() here as that can require a ResourceMark which we
  // can't use because this gets called too early in the thread initialization.
  assert(_stack_base != nullptr, "Stack base not yet set for thread id:%d (0 if not set)",
         osthread() != nullptr ? osthread()->thread_id() : 0);
}
#endif
```

File: src/hotspot/share/runtime/thread.hpp
Comment (shipilev) [inline src/hotspot/share/runtime/thread.hpp:535]: I think it is a bit cleaner to "just" outline the checking method like:

```
thread.hpp:
  address stack_base() const           { assert_stack_base(); return _stack_base; }
  void assert_stack_base() const NOT_DEBUG_RETURN;

thread.cpp:
#ifdef ASSERT
void  Thread::assert_stack_base() const {
  // Note: can't report Thread::name() here as that can require a ResourceMark which we
  // can't use because this gets called too early in the thread initialization.
  assert(_stack_base != nullptr, "Stack base not yet set for thread id:%d (0 if not set)",
         osthread() != nullptr ? osthread()->thread_id() : 0);
}
#endif
```

**Review by shipilev — COMMENTED** [review]
Nitpicking...

File: src/hotspot/share/runtime/thread.hpp
Comment (shipilev) [reply src/hotspot/share/runtime/thread.hpp:535]: This _is_ the way we do it in existing code, e.g.
https://github.com/openjdk/jdk/blob/46b02f49bcc730d94e37cf17fa996fdd12bdb990/src/hotspot/share/runtime/stackWatermark.hpp#L105

Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:535]: So many ways to achieve the same thing ...

I would not want assert_stack_base in the public API.

I did need to fix the ifdef though.

Comment (shipilev) [reply src/hotspot/share/runtime/thread.hpp:535]: `DEBUG_ONLY(;)` just triggers me, sorry. We have `NOT_DEBUG_RETURN` exactly for this reason: mark the methods that are only defined in debug builds, and shortcut otherwise :)

**Comment by stefank** [comment]
> Thanks for the review @stefank !
> 
> > If it is crashing when called on another thread, it might be easier to add informative asserts in that sub-system instead. You could for example add a non-asserting getter and perform all relevant checks on the caller side.
> 
> The assert is firing when `stack_base()` is called on another thread. In that case we had nothing to tell us which thread it was, hence this enhancement to provide that information.

That was fully understood and I think you are missing my point. I was trying to convey that it might be better to rewrite the calling code to perform its own checks / logging instead of relying on a "low-level" assert in Thread. 2c.

**Comment by dholmes-ora** [comment]
> I was trying to convey that it might be better to rewrite the calling code to perform its own checks / logging instead of relying on a "low-level" assert in Thread. 

The calling code is of course being fixed (or at least the cause dealt with one way or another). The point here was to simply enhance the existing assertion to be more informative for the next time it gets triggered.

**Comment by stefank** [comment]
> > I was trying to convey that it might be better to rewrite the calling code to perform its own checks / logging instead of relying on a "low-level" assert in Thread.
> 
> The calling code is of course being fixed (or at least the cause dealt with one way or another). The point here was to simply enhance the existing assertion to be more informative for the next time it gets triggered.

Right. And that was what I was curious to know more about because I think there would be better ways to get better diagnostics by changing code there. Anyways, I'll leave you to deal with that in a separate issue.

**Comment by dholmes-ora** [comment]
Anyone want to add the second review or shall I just abort this?

**Review by kevinjwalls — APPROVED** [review]
Looks good, and useful!

**Review by shipilev — APPROVED** [review]
Well, I am still nitpicking about `DEBUG_ONLY(;)`, but I can live with it.

**Comment by stefank** [comment]
This PR is much more complicated than it needs to be.

1) We would probably be fine with always having the implementation in the .cpp file. I don't think that we need to optimize away the call for this getter. I don't see hot code using this function.

2) If we really want to optimize it away, we can just stick the assert in the header just like we already do with related functions:

```
  bool is_in_stack_range(address adr, address limit, bool inclusive) const {
    assert(stack_base() > limit && limit >= stack_end(), "limit is outside of stack");
    return stack_base() > adr && (inclusive ? adr >= limit : adr > limit);
  }
```

There's no point in trying to separate the debug/release builds into .cpp/.hpp files if the related functions already put the assert in the header.

**Comment by dholmes-ora** [comment]
>  we can just stick the assert in the header just like we already do with related functions:

@stefank  as mentioned in the description I ran into include file issues referring to `osthread()->thread_id()` in thread.hpp, as we only have a forward decl of `OSThread`.

**Comment by dholmes-ora** [comment]
Thanks for the review @kevinjwalls !

**Comment by stefank** [comment]
> > we can just stick the assert in the header just like we already do with related functions:
> 
> @stefank as mentioned in the description I ran into include file issues referring to `osthread()->thread_id()` in thread.hpp, as we only have a forward decl of `OSThread`.

OK. Did you consider my bullet (1) at all?

**Comment by stefank** [comment]
> > > we can just stick the assert in the header just like we already do with related functions:
> > 
> > 
> > @stefank as mentioned in the description I ran into include file issues referring to `osthread()->thread_id()` in thread.hpp, as we only have a forward decl of `OSThread`.
> 
> OK. Did you consider my bullet (1) at all?

I re-read the PR and see that this is what you started with and that Thomas requested this to be inlined.

**Comment by stefank** [comment]
FTR, I'm not blocking this, I'm just a little annoyed that we unnecessarily (IMHO) uglify the code.

**Comment by dholmes-ora** [comment]
Okay thanks for the reviews.

/integrate

**Pull request closed without merging**
